File: app/src/main/java/net/yanzm/behaviorsubjectsample/MainActivity.java
package net.yanzm.behaviorsubjectsample;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.schedulers.Schedulers;
import rx.subjects.BehaviorSubject;
import rx.subscriptions.Subscriptions;

public class MainActivity extends AppCompatActivity implements MainFragment.MainFragmentListener {

    private final BehaviorSubject<CommonData> commonDataBehaviorSubject = BehaviorSubject.create();

    @NonNull
    private Subscription subscription = Subscriptions.empty();

    @BindView(R.id.swipe_refresh_layout)
    SwipeRefreshLayout swipeRefreshLayout;

    @BindView(R.id.pager)
    ViewPager viewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refresh();
            }
        });

        viewPager.setAdapter(new MyPagerAdapter(getSupportFragmentManager()));

        refreshCommonData();
    }

    @Override
    protected void onDestroy() {
        commonDataBehaviorSubject.onCompleted();
        subscription.unsubscribe();
        super.onDestroy();
    }

    private void refresh() {
        refreshCommonData();

        // 現在インスタンス化されている MainFragment に取り直しを命じる
        final List<Fragment> fragments = getSupportFragmentManager().getFragments();
        if (fragments != null) {
            for (Fragment fragment : fragments) {
                if (fragment instanceof MainFragment) {
                    ((MainFragment) fragment).refresh();
                }
            }
        }
    }

    private void refreshCommonData() {
        // 取り直しなのでここで null をセットして新しく作られたタブで前の値が流れないようにする
        commonDataBehaviorSubject.onNext(null);

        subscription = DataRetriever.getInstance().getCommonData()
                .onErrorReturn(new Func1<Throwable, CommonData>() {
                    @Override
                    public CommonData call(Throwable throwable) {
                        // エラーのときはデータがないものとして扱う
                        return CommonData.empty();
                    }
                })
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Action1<CommonData>() {
                    @Override
                    public void call(CommonData commonData) {
                        swipeRefreshLayout.setRefreshing(false);
                        commonDataBehaviorSubject.onNext(commonData);
                    }
                });
    }

    @NonNull
    @Override
    public Observable<CommonData> getCommonDataObservable() {
        return commonDataBehaviorSubject
                .filter(new Func1<CommonData, Boolean>() {
                    @Override
                    public Boolean call(CommonData commonData) {
                        // 取り直しのときに以前の値を流さないため
                        return commonData != null;
                    }
                });
    }

    static class MyPagerAdapter extends FragmentPagerAdapter {

        MyPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int position) {
            return MainFragment.newInstance(position);
        }

        @Override
        public int getCount() {
            return 10;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return "TAB" + position;
        }
    }

}

File: app/src/main/java/net/yanzm/behaviorsubjectsample/rx/MainActivity.kt
package net.yanzm.behaviorsubjectsample

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import net.yanzm.behaviorsubjectsample.CommonData.Companion.empty
import net.yanzm.behaviorsubjectsample.DataRetriever.getCommonData
import net.yanzm.behaviorsubjectsample.databinding.ActivityMainBinding
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subjects.BehaviorSubject
import rx.subscriptions.Subscriptions

class MainActivity : AppCompatActivity(), MainFragmentListener {

    private val commonDataBehaviorSubject = BehaviorSubject.create<CommonData?>()
    private var subscription = Subscriptions.empty()
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.swipeRefreshLayout.setOnRefreshListener { refresh() }
        binding.viewPager.adapter = MyPagerAdapter(supportFragmentManager)

        refreshCommonData()
    }

    override fun onDestroy() {
        commonDataBehaviorSubject.onCompleted()
        subscription.unsubscribe()
        super.onDestroy()
    }

    private fun refresh() {
        refreshCommonData()

        // 現在インスタンス化されている MainFragment に取り直しを命じる
        for (fragment in supportFragmentManager.fragments) {
            if (fragment is MainFragment) {
                fragment.refresh()
            }
        }
    }

    private fun refreshCommonData() {
        // 取り直しなのでここで null をセットして新しく作られたタブで前の値が流れないようにする
        commonDataBehaviorSubject.onNext(null)

        subscription = getCommonData()
            .onErrorReturn { empty() }
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                binding.swipeRefreshLayout.isRefreshing = false
                commonDataBehaviorSubject.onNext(it)
            }
    }

    override fun getCommonDataObservable(): Observable<CommonData> {
        return commonDataBehaviorSubject.filter { it != null }
    }
}

private class MyPagerAdapter(fm: FragmentManager) :
    FragmentPagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

    override fun getItem(position: Int): Fragment = MainFragment.newInstance(position)

    override fun getCount(): Int = 10

    override fun getPageTitle(position: Int): CharSequence = "TAB$position"
}
